File: lista2/lista2.ml
(* (function x -> x+x)6;;
let double = function x -> x+x;;

let f = function y -> function x -> x*x +y;;

let plus (x,y) = x+y;;

let rec succ_tail n = 
  let rec succ_iter(n, acc) = 
    if n = 0 then acc
    else succ_iter(n-1, acc+1)
  in succ_iter(n,1);;

let xs = ["a";"b";"c"];;
(* h -> "a"
t -> ["b";"c"] *)

let (z, _) = (false, 10);;

let x = (("SMith", 25), true);;
let ((n,w),b) = x;;
(* n smith, w 25, b bool *)

let imply1 pb = 
  match pb with
    (false, false) -> true
  | (false, true) -> true
  | (true, false) -> false
  | (true, true) -> true
;;


let rec zip(xs, ys) =
  match (xs, ys) with
  (h1::t1, h2::t2) -> (h1, h2)::zip(t1,t2)
  | _ -> [];;
 
  (* [(1,'a');(2,'b');(3,'c')] *)
  (* 1::[2;3];; -> [1;2;3];; *)
zip([1;2;3], ['a';'b';'c']);; 

let rec unzip ps =
  match ps with 
  [] -> ([], [])
  | (h1,h2)::t -> let (l1, l2) = unzip t in (h1::l1, h2::l2);;

let isLatin v = 
  match v with 
  'a' | 'e' | 'i' | 'o' | 'u' -> true
  | _ -> false;;


let avg p =
  match p with
  (x,y) when x=y -> y
  | (x,y) -> (x +. y) /. 2.0;;

 avg(5., 5.);; *)

(* 
 let rec sqr_list li = 
  match li with
  [] -> []
  | h:: tail -> h*h :: sqr_list t;; *)


(* let isPrime x =
  match x with
  0 -> false
  | 1 -> false
  | _ -> let rec primerHelper(n,i) =
    match i with
    1 -> true
    | _ -> if n mod i = 0 then false else primerHelper(n,i-1) in
    primerHelper(x, x-1);;

let res = isPrime 4;;
Printf.printf "%B" res;; *)

let isPrime x = 
  let rec primeHelper(x,acc) =
    if x < 2 then false
    else if x = 2 then true
    else if x mod acc = 0 then false
    else if acc*acc > x then true
    else primeHelper(x, acc+1)
  in primeHelper(x, 2);;

let res = isPrime 8;;
Printf.printf "task 8 result: %B" res;;

let evenIndexList lis = 
  let rec evenIndexListHelper(arr, resultList, position) =
    if arr = [] then resultList
    else if position mod 2 = 0 then evenIndexListHelper(List.tl arr, resultList@[List.hd arr], position +1)
    else evenIndexListHelper(List.tl arr, resultList, position +1)
  in evenIndexListHelper(lis, [], 1);;
evenIndexList([1;2;3;4;5;6;7]);;

let rec existInArr (x, arr) =
  match arr with 
    [] -> false
    | h::t -> if h = x then true else existInArr(x, t);;

let removeDuplicates lis =
  let rec removeDuplicates(arr, resultList) = 
    if arr = [] then resultList
    else if existInArr(List.hd arr, List.tl arr) then removeDuplicates(List.tl arr, resultList)
    else removeDuplicates(List.tl arr, resultList@[List.hd arr])
  in removeDuplicates(lis, []);;

removeDuplicates([1;2;3;1;4;6;3]);;

let palindrome xs =
  let rec palindrome_helper(arr, acc)=
  if arr = [] then acc = xs
  else palindrome_helper(List.tl arr, List.hd arr::acc)
in palindrome_helper(xs, []);;

palindrome([1;2;1;]);;

let reverse xs =
  let rec reverse_helper(arr, acc) = 
    if arr = [] then acc
    else reverse_helper(List.tl arr, List.hd arr::acc)
  in reverse_helper(xs, []);;

reverse([1;2;3;4]);;


let length xs =
  let rec length_helper(arr, acc) = 
    if arr = [] then acc
    else length_helper(List.tl arr, acc+1)
  in length_helper(xs,0);;

length([1;2;3;4;5]);;

let rec two_el xs =
  match xs with
  | [] -> None
  | [x] -> None
  | [x;y] -> Some(x,y)
  | h::t -> two_el t;;
two_el([1;2;3;4;5]);;

let rec one_el xs = 
  match xs with
  | []  -> None
  | [x] -> Some(x)
  | h::t -> one_el t;;
one_el([1;2;3;4;5]);;

File: lista2/src/main/scala/Lista2.scala
import scala.annotation.tailrec

object Lista2 extends App {

  println("lista2")
  ((x:Int) => x+x)(6) // f anonimowa
  val varFunc = ((x:Int) => x+x)(6) // f anonimowa
  val f = (y:Int) => (x:Int) => x*x+y;
  f(2)(5)

  def succTail(n: Int): Int = {
    def succIter(n:Int, acc: Int): Int = {
      if(n==0) acc
      else succIter(n-1, acc+1);
    }
    succIter(n,1)
  }

  val xs = List("a", "b", "c");
  val List(x1,x2,x3) = xs
  val h::t = xs;
  // h -> a
  // t -> ["b", "c"]

  val (z,_) = (false, 10); //wildcard

  val x = (("smith", 25), true);
  val ((a,b),c) = x

  //match pattern
  def imply1(pb:(Boolean, Boolean)): Boolean = {
    pb match {
      case (false, false) => true
      case (false, true) => true
      case (true, false) => false
      case (true, true) => true
    }

    imply1(1>2, true)
  }

  def zip[A,B](xs: List[A], ys: List[B]): List[(A,B)] =
    (xs, ys) match {
      case (h1::t1, h2::t2) => (h1, h2)::zip(t1,t2)
      case _ => Nil
    }
  zip(List(1,2,3), List('a', 'b', 'c')) // List((1,'a'), (2, 'b'), (3, 'c'))



  def unzip[A,B](ps: List[(A,B)]): (List[A], List[B]) =
    ps match {
      case Nil => (Nil, Nil)
      //h1 -> 1
      //h2 -> 2
      //t -> List((3,4), (5,6), (7,8))
      case (h1, h2)::t => {val(xs1, xs2) = unzip(t); (h1::xs1, h2::xs2)}
    }

  unzip(List((1,2), (3,4), (5,6), (7,8))) //(List(1,3,5,7), List(2,4,6,8))

  def avg (p:(Double, Double)) =
    p match {
      case (x,y) if x == y => x
      case (x, y) => (x+y) /2.0
    }


  def isPrime(x:Int) = isPrimeHelper(x, 2)
  @tailrec
  def isPrimeHelper(i: Int, acc: Int): Boolean={
    if (i < 2) false
    else if (i == 2) true
    else if (i % acc == 0) false
    else if (acc*acc > i) true
    else isPrimeHelper(i, acc+1)
  }
  println(isPrime(4))

  def evenIndexList[A](xs: List[A]) = evenHelper(xs, List(), 1);
  @tailrec
  def evenHelper[A](arr: List[A], result:List[A], pos: Int): List[A] = {
    if (arr == List()) result
    else if(pos % 2 == 0) evenHelper(arr.tail, result.appended(arr.head), pos+1)
    else evenHelper(arr.tail, result, pos+1);
  }
  print(evenIndexList(List(1,2,3,4,5,6)))


  @tailrec
  def exists[A](num: A, arr: List[A]):Boolean= {
    arr match {
      case Nil => false
      case (h::t) => if (h==num) true else exists(num, t)
    }
  }
  def remDup[A](xs: List[A]):List[A] = remDupHelperList(xs, List())
  @tailrec
  def remDupHelperList[A](arr: List[A], res: List[A]): List[A] = {
    if (arr == List()) res
    else if(exists(arr.head, arr.tail)) remDupHelperList(arr.tail, res)
    else remDupHelperList(arr.tail, res ::: List(arr.head))
  }
  println(remDup(List(1,2,2,3,4,5,5)))

  def pali[A](xs: List[A]):Boolean = if(reverseHelper(xs, List()) == xs) true else false
  def reverse[A](xs:List[A]): List[A] = reverseHelper(xs, List());
  @tailrec
  def reverseHelper[A](arr: List[A], result: List[A]): List[A] = {
    if (arr == List()) result
    else reverseHelper(arr.slice(0, arr.length-1), result.appended(arr.last))
  }
  println(pali(List(1,2,1)))
  println(reverse(List(1,2,2)))

  def length[A](xs:List[A]): Int = lengthHelper(xs, 0)
  @tailrec
  def lengthHelper[A](arr: List[A], count: Int): Int = {
    if (arr == List()) count
    else lengthHelper(arr.tail, count+1)
  }

  println(length(List(1,2,3,4,5)))

  @tailrec
  def twoLast[A](xs: List[A]): Option[(A,A)] = {
    xs match {
      case Nil => None
      case h :: Nil => None
      case h::t:: Nil => Some(h,t)
      case h::t => twoLast(t)
    }
  }

  // stworzyc kazda funckja jako osoby program i obsluzyc wejscie i wyjscie stanradowe wprowadz input
  println(twoLast(List(1,2,4,5,4,5)))

  def oneLast[A](xs: List[A]): Option[A] = {
    xs match {
      case Nil => None
      case h:: Nil => Some(h)
      case h::t => oneLast(t)
    }
  }
  println(oneLast(List(1,2,3,4,5,6)))

}

